nova/virt/vmwareapi/vmops.py

```diff
@@ -2232,3 +2247,88 @@ def get_mks_console(self, instance):
+        vim = self._session.vim
+        options = None
+        if self._property_collector is None:
+            self._property_collector = vim.service_content.propertyCollector
+            vim.CreateFilter(self._property_collector, spec=self._get_vm_monitor_spec(vim), partialUpdates=False)
+        update_set = vim.WaitForUpdatesEx(self._property_collector, version="",
+                                          options=options)
+        while update_set:
+
+            self._property_collector_version = update_set.version
+            if update_set.filterSet and update_set.filterSet[0].objectSet:
+                for update in update_set.filterSet[0].objectSet:
+                    if update.obj['_type'] == "VirtualMachine":
+                        if update.kind == "leave":
+                            LOG.info("Removing instance from cache...")
+                            vm_refs_cache = copy.copy(vm_util._VM_REFS_CACHE)
+                            inv_cache_map = {v.value: k for k, v in vm_refs_cache.items()}
+                            cache_id_to_delete = inv_cache_map.get(update.obj.value)
+                            vm_util.vm_ref_cache_delete(cache_id_to_delete)
+                            vm_util.vm_value_cache_delete(cache_id_to_delete)
+                        else:
+                            if update.kind == "enter":
+                                for key, change in enumerate(update.changeSet):
+                                    if key == 0 or key == 1:
+                                        continue
+                                    else:
+                                        if change['op'] == 'assign':
+                                            if hasattr(update.changeSet[0], 'val') and hasattr(update.changeSet[1],'val'):
+                                                vm_util._VM_VALUE_CACHE[update.changeSet[0].val][change.name] = change.val
+                                if hasattr(update.changeSet[0], 'val') and hasattr(update.changeSet[1], 'val'):
+                                    if update.changeSet[1].val.extensionKey == constants.EXTENSION_KEY:
+                                        LOG.info("Adding vm to cache...")
+                                        vm_util.vm_ref_cache_update(update.changeSet[0].val, update.obj)
+
+                            elif update.kind == "modify":
+                                if update.changeSet[0].name == "config.instanceUuid":
+                                    if update.changeSet[1].val.extensionKey == constants.EXTENSION_KEY:
+                                        vm_util.vm_ref_cache_update(update.changeSet[0].val, update.obj)
+                                else:
+                                    vm_refs_cache = copy.copy(vm_util._VM_REFS_CACHE)
+                                    inv_cache_map = {v.value: k for k, v in vm_refs_cache.items()}
+                                    cache_id_to_update = inv_cache_map.get(update.obj.value)
```

Comment:
Here you could drop the `inv_cache_map` and set `cache_id_to_update = update.obj.value`